src/v/cluster/types.h

```diff
@@ -1312,6 +1315,31 @@ struct topic_configuration
         return std::tie(tp_ns, partition_count, replication_factor, properties);
     }
 
+    void serde_read(iobuf_parser& in, const serde::header& h) {
+        using serde::read_nested;
+
+        tp_ns = read_nested<model::topic_namespace>(in, h._bytes_left_limit);
+        partition_count = read_nested<int32_t>(in, h._bytes_left_limit);
+        replication_factor = read_nested<int16_t>(in, h._bytes_left_limit);
+        properties = read_nested<topic_properties>(in, h._bytes_left_limit);
+
+        if (h._version < 1) {
+            maybe_adjust_retention_policies(
```

Comment:
should we also adjust retention policies of `topic_configurations` serialized with ADL ? older redpanda versions (even after upgrading them step by step) may still have old version of topic configuration serialized with ADL in controller log

Reply:
I didn't do it because v22.2 uses serde for controller messages. This, plus the fact that create topic requests are not forwarded if they hit a follower, means that this type should never be encoded in ADL.

Reply:
Since controller log is set in stone and it's content is not migrated from the ADL when running new redpanda version we must create similar code path for adl serialized controller commands. When a cluster was operational before we introduced serde serialization for controller commands all of them were serialized with ADL. Now when replaying the controller log during startup redpanda will read those ADL serialized commands and they will not be adjusted with the new settings

Reply:
I see. So we need it for the following scenario:
A node running v22.1 (last pre-serde for controller log version) persists a `topic_configuration`. Once we finally upgrade to v22.3 that message is still encoded in ADL and we need to know how to handle it.

Sounds reasonable. Will add it.

Reply:
Done.